javascript/ql/src/Security/CWE-078/CommandInjection.inc.qhelp

```diff
@@ -2,43 +2,63 @@
   "-//Semmle//qhelp//EN"
   "qhelp.dtd">
 <qhelp>
-<overview>
-<p>Code that passes user input directly to
-<code>require('child_process').exec</code>, or some other library
-routine that executes a command, allows the user to execute malicious
-code.</p>
 
+<overview>
+<p>Code that passes untrusted user input directly to
+<code>child_process.exec</code> or similar APIs that execute commands by
+spawning a shell allows the user to execute malicious code.</p>
 </overview>
+
 <recommendation>
+<p>If possible, use hard-coded string literals to specify the command to run.
+Instead of interpreting the user input directly as a shell command, examine the
+user input and then choose among hard-coded string literals.</p>
+
+<p>If the applicable libraries or commands cannot be determined until runtime,
+then add code to verify that the user input string is safe before using it.</p>
 
-<p>If possible, use hard-coded string literals to specify the command to run
-or library to load. Instead of passing the user input directly to the
-process or library function, examine the user input and then choose
-among hard-coded string literals.</p>
+<p>If possible, prefer APIs that run the commands directly rather than via a
+shell, and that accept command arguments as an array of strings rather than a
+single concatenated string. This is both safer and more portable.</p>
```

Comment:
```suggestion
<p>If possible, use APIs that don't run shell commands, and accept command arguments
as an array of strings rather than a single concatenated string. This is both safer and more portable.</p>
```